crudd/src/main.rs

```diff
@@ -226,19 +225,89 @@ fn cmd_read<T: BlockIO>(opt: &Opt, crucible: Arc<T>) -> Result<()> {
         let buffer = Buffer::new((blocks * native_block_size) as usize);
         let block_idx = (cmd_count * opt.iocmd_block_count) + block_offset;
         let offset = Block::new(block_idx, native_block_size.trailing_zeros());
-        let mut waiter = crucible.read(offset, buffer.clone())?;
-        waiter.block_wait()?;
-        output.write_all(&buffer.as_vec()[0..remainder as usize])?;
+        let waiter = crucible.read(offset, buffer.clone()).await?;
+        waiter.wait().await?;
+        output.write_all(&buffer.as_vec().await[0..remainder as usize])?;
     }
 
     Ok(())
 }
 
-fn cmd_write<T: BlockIO>(opt: &Opt, crucible: Arc<T>) -> Result<()> {
+// We need to run this at the end to handle the remainder bytes, but we also
```

Comment:
this was perviously a function i defined inside cmd_write to close over a bunch of variables calculater before the write loop gets going. my assumption is async makes that kind of closure not work. all the captured variables would need to be moved into the function since the borrow checker can't guarantee they'll live as long as the async function, whereas previously they were read-only borrowed and everything works out fine. so that's probably why it got moved out into its own thing.